go/tools/go-upgrade/go_upgrade.go

```diff
@@ -0,0 +1,453 @@
+	if err != nil {
+		return nil, err
+	}
+	content := string(contentRaw)
+	idxBegin := strings.Index(content, "goversion_min") + len("goversion_min") + 1
+	idxFinish := strings.Index(content[idxBegin:], "||") + idxBegin
+	versionStr := strings.TrimSpace(content[idxBegin:idxFinish])
+	return version.NewVersion(versionStr)
+}
+
+func currentBootstrapVersion() (float64, error) {
+	contentRaw, err := os.ReadFile("Makefile")
+	if err != nil {
+		return 0, err
+	}
+	content := string(contentRaw)
+	idxBegin := strings.Index(content, "BOOTSTRAP_VERSION=") + len("BOOTSTRAP_VERSION=")
+	idxFinish := strings.IndexByte(content[idxBegin:], '\n') + idxBegin
+	versionStr := strings.TrimSpace(content[idxBegin:idxFinish])
+	f, err := strconv.ParseFloat(versionStr, 64)
+	if err != nil {
+		return 0, err
+	}
+	return f, nil
+}
+
+// getLatestStableGolangReleases fetches the latest stable releases of Golang from
+// the official website using the goDevAPI URL.
+// Once fetched, the releases are returned as version.Collection.
+func getLatestStableGolangReleases() (version.Collection, error) {
+	resp, err := http.Get(goDevAPI)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var latestGoReleases []latestGolangRelease
+	err = json.Unmarshal(body, &latestGoReleases)
+	if err != nil {
+		return nil, err
+	}
+
+	var versions version.Collection
+	for _, release := range latestGoReleases {
+		if !release.Stable {
+			continue
+		}
+		if !strings.HasPrefix(release.Version, "go") {
+			return nil, fmt.Errorf("golang version malformatted: %s", release.Version)
+		}
+		newVersion, err := version.NewVersion(release.Version[2:])
+		if err != nil {
+			return nil, err
+		}
+		versions = append(versions, newVersion)
+	}
+	return versions, nil
+}
+
+// chooseNewVersion decides what will be the next version we're going to use in our codebase.
+// Given the current Golang version, the available latest versions and whether we allow major upgrade or not,
+// chooseNewVersion will return either the new version or nil if we cannot/don't need to upgrade.
+func chooseNewVersion(curVersion *version.Version, latestVersions version.Collection, allowMajorUpgrade bool) *version.Version {
+	selectedVersion := curVersion
+	for _, latestVersion := range latestVersions {
+		if !allowMajorUpgrade && !isSameMajorVersion(latestVersion, selectedVersion) {
+			continue
+		}
+		if latestVersion.GreaterThan(selectedVersion) {
+			selectedVersion = latestVersion
+		}
+	}
+	// No change detected, return nil meaning that we do not want to have a new Golang version.
+	if selectedVersion.Equal(curVersion) {
+		return nil
+	}
+	return selectedVersion
+}
+
+// replaceGoVersionInCodebase goes through all the files in the codebase where the
+// Golang version must be updated
+func replaceGoVersionInCodebase(old, new *version.Version, noWorkflowUpdate bool) error {
+	explore := []string{
```

Comment:
IMO it's worth doing a check here: if old and new are the same then return.

Reply:
This is done in the caller of that method:

```go
	if upgradeTo == nil {
		return nil
	}
```

Reply:
Added an extra safety net via 223e72d72a :)